File: src/WakePOD.jl
include("SVD.jl")
using NPZ
using GLMakie
using LinearAlgebra
using Statistics

xlims, ylims = (-0.35, 2), (-0.35, 0.35)

flow_snapshots_u = npzread("data/u.npy")
flow_snapshots_u = permutedims(flow_snapshots_u, [3, 2, 1])
flow_snapshots_v = npzread("data/v.npy")
flow_snapshots_v = permutedims(flow_snapshots_v, [3, 2, 1])
# Just the wake
nx, ny, nt = size(flow_snapshots_u)
dt = 8/nt
pxs = LinRange(xlims..., nx)
pys = LinRange(ylims..., ny)

wake_snapshots_u = flow_snapshots_u[pxs.>1, :, :]
wake_snapshots_v = flow_snapshots_v[pxs.>1, :, :]
pxs = pxs[pxs.>1]
nx, ny, nt = size(wake_snapshots_u)

# take away the time mean
# flow_snapshots_u = flow_snapshots_u .- mean(flow_snapshots_u, dims = 3)
# flow_snapshots_v = flow_snapshots_v .- mean(flow_snapshots_v, dims = 3)

# get mag and vorticity
function curl(u, v, dx, dy)
    ∂v_∂x = zeros(size(v))
    ∂u_∂y = zeros(size(u))
    
    # Compute ∂v/∂x
    @views ∂v_∂x[:, 2:end-1] .= (v[:, 3:end] - v[:, 1:end-2]) / (2 * dx)
    
    # Compute ∂u/∂y
    @views ∂u_∂y[2:end-1, :] .= (u[3:end, :] - u[1:end-2, :]) / (2 * dy)
    
    return ∂v_∂x - ∂u_∂y
end

# Calculate vorticity snapshots
vorticity_snapshots = zeros(size(wake_snapshots_u))

dx = mean(pxs[2:end] - pxs[1:end-1])
dy = mean(pys[2:end] - pys[1:end-1])

for t in 1:nt
    vorticity_snapshots[:, :, t] = curl(wake_snapshots_u[:, :, t], wake_snapshots_v[:, :, t], dx, dy)
end

function testplotvort()
    # Quick test plot
    f = Figure(resolution = (1600, 400))
    ax = Axis(f[1, 1])
    co = contourf!(ax,
        pxs, pys,vorticity_snapshots[:,:,1],
        xlabel=L"x", ylabel=L"y", title=L"ω_z",
        levels = range(-0.05, 0.05, length = 44),
        colormap=:seismic,
        extendlow = :auto, extendhigh = :auto,
    )
    tightlimits!(ax)
    return f
end
f = testplotvort()
save("figures/testplotvort.png", f)

mag_snapshots = (wake_snapshots_u.^2 .+ wake_snapshots_v.^2).^0.5

function testplotmag()
    # Quick test plot
    f = Figure(resolution = (1600, 400))
    ax = Axis(f[1, 1])
    co = contourf!(ax,
        pxs, pys,mag_snapshots[:,:,1],
        xlabel=L"x", ylabel=L"y", title=L"ω_z",
        levels = range(-0.1, 1.2, length = 44),
        # colormap=:icefire,
        extendlow = :auto, extendhigh = :auto,
    )
    tightlimits!(ax)
    return f
end
f = testplotmag()
save("figures/testplotmag.png", f)


# Flatten snapshots along first two dims
flat_snapshots_u = reshape(
    wake_snapshots_u,
    nx*ny,
    nt
    )
flat_snapshots_v = reshape(
    wake_snapshots_v,
    nx*ny,
    nt
    )
flat_vorticity_snapshots = reshape(
    vorticity_snapshots,
    nx*ny,
    nt
    )

k=20
U, Σ, Vk = rSVD(flat_vorticity_snapshots, k)

# for m in 1:k
#     f = Figure(resolution = (1600, 400))
#     ax = Axis(f[1, 1])
#     ϕ = reshape(U[:, m], nx, ny)
#     println(minimum(ϕ), " ", maximum(ϕ))
#     co = contourf!(ax,
#         pxs, pys, reshape(U[:, m], nx, ny),
#         xlabel=L"x", ylabel=L"y", title=L"ϕ_$m",
#         levels = range(-0.01, 0.01, length = 44),
#         # colormap=:plasma,
#         extendlow = :auto, extendhigh = :auto,
#     )
#     tightlimits!(ax)
#     save("figures/modes/U_$m.png", f)
# end

A = diagm(Σ[1:k]) * Vk

using FileIO
using ImageMagick

# Loop over each mode
for m in 1:k
    # Create an array to store frames for the GIF
    Mk = U[:,m:m] * A[m:m, :]
    frames = []

    # Loop over each time step
    for t in 1:nt
        Mpl = reshape(Mk[:, t], nx, ny)
        fig = Figure(resolution = (1600, 400))
        ax = Axis(fig[1, 1])
        lines = contourf!(ax, pxs, pys, Mpl,
            levels = range(-0.01, 0.01, length = 44),
            colormap = :seismic,
            extend = :both,
        )
        xlabel!(ax, "x")
        ylabel!(ax, "y")
        title!(ax, "ϕ_$(m)(t)")
        tightlimits!(ax)

        # Push the current frame to the frames array
        push!(frames, frame(fig))
    end

    # Save the frames as a GIF
    save(string("figures/time/mode_", m, ".gif"), frames, fps = 10)
end

File: src/WakePOD.py
import numpy as np
from scipy.linalg import qr, svd
import matplotlib.pyplot as plt
import seaborn as sns
import scienceplots
plt.style.use(["science"])
plt.rcParams["font.size"] = "10.5"

xlims, ylims = (1, 2), (-0.35, 0.35)

vort = np.load("data/wake_vort.npy")

nx, ny, nt = vort.shape
print("Loaded snaps")
# Just the wake
dt = 8 / nt
pxs = np.linspace(*xlims, nx)
pys = np.linspace(*ylims, ny)


# Flatten snapshots along first two dimensions
flat_vort = vort.reshape(nx * ny, nt)

def rSVD(A, k):
    m, n = A.shape
    Φ = np.random.rand(n, k)
    Ar = A @ Φ
    Q, _ = qr(Ar, mode='economic')
    B = Q.T @ A
    U, S, Vt = svd(B, full_matrices=False)
    max_idx = np.argsort(S)[::-1][:k]
    Sk = S[max_idx]
    Uk = U[:, max_idx]
    Vk = Vt[max_idx, :]
    U = Q @ Uk
    return U, Sk, Vk


k = 20
U, S, Vk = rSVD(flat_vort, k)

np.save("U_POD.npy", U)
np.save("S_POD.npy", S)
np.save("Vk_POD.npy", Vk)


for m in range(k):
    fig, ax = plt.subplots(figsize=(5, 3))
    lim = [-.005, .005]
    levels = np.linspace(lim[0], lim[1], 44)
    _cmap = sns.color_palette("seismic", as_cmap=True)
    phi = U[:, m].reshape(nx, ny)
    print(np.min(phi), np.max(phi))
    co = ax.contourf(
        pxs,
        pys,
        phi.T,
        levels=levels,
        vmin=lim[0],
        vmax=lim[1],
        # norm=norm,
        cmap=_cmap,
        extend="both",
    )
    ax.set_aspect(1)
    ax.set(xlabel=r"$x$", ylabel=r"$y$", title=r"$\phi_{" + str(m) + r"}$")
    # fig.colorbar(co)
    plt.tight_layout()
    fig.savefig("figures/modes/U_{}.png".format(m))

# A = np.diag(S[:k]) @ Vk[:k, :]
# M1 = U[:, :1] @ A[:1, :]

# # Animate the modes
# for m in range(k):
#     Mk = U[:, m:m+1] @ A[m:m+1, :]
#     for t in range(nt):
#         fig = plt.figure(figsize=(16, 4))
#         ax = fig.add_subplot(1, 1, 1)
#         co = ax.contourf(
#             pxs, pys, Mk[:, t].reshape(nx, ny),
#             levels=np.linspace(-0.01, 0.01, num=44),
#             cmap='seismic',
#             extend='both'
#         )
#         ax.set(xlabel=r"$x$", ylabel=r"$y$", title=r"$\phi_1(t)$")
#         fig.colorbar(co)
#         plt.tight_layout()
#         fig.savefig("figures/time/M{}_{}.png".format(m, t))

# # # Now let's get the DMD
# Phi = U @ np.sqrt(np.diag(S)) @ Vk  # DMD modes
# print(Phi.shape)
# Lambda = np.sqrt(np.diag(S)) @ Vk  # Temporal dynamics
# print(vort.shape)